packages/api-fetch/src/index.js

```diff
@@ -38,106 +38,117 @@ const DEFAULT_OPTIONS = {
+		url || path,
+		{
+			...DEFAULT_OPTIONS,
+			...remainingOptions,
+			body,
+			headers,
+		}
+	);
+	const checkStatus = ( response ) => {
+		if ( response.status >= 200 && response.status < 300 ) {
+			return response;
+		}
 
-		// Merge explicitly-provided headers with default values.
-		headers = { ...DEFAULT_HEADERS, ...headers };
+		throw response;
+	};
 
-		// The `data` property is a shorthand for sending a JSON body.
-		if ( data ) {
-			body = JSON.stringify( data );
-			headers[ 'Content-Type' ] = 'application/json';
+	const parseResponse = ( response ) => {
+		if ( parse ) {
+			if ( response.status === 204 ) {
+				return null;
+			}
+
+			return response.json ? response.json() : Promise.reject( response );
 		}
 
-		const responsePromise = window.fetch(
-			url || path,
-			{
-				...DEFAULT_OPTIONS,
-				...remainingOptions,
-				body,
-				headers,
-			}
-		);
-		const checkStatus = ( response ) => {
-			if ( response.status >= 200 && response.status < 300 ) {
-				return response;
-			}
+		return response;
+	};
 
-			throw response;
-		};
+	return responsePromise
+		.then( checkStatus )
+		.then( parseResponse )
+		.catch( ( response ) => {
+			if ( ! parse ) {
+				throw response;
+			}
 
-		const parseResponse = ( response ) => {
-			if ( parse ) {
-				if ( response.status === 204 ) {
-					return null;
-				}
+			const invalidJsonError = {
+				code: 'invalid_json',
+				message: __( 'The response is not a valid JSON response.' ),
+			};
 
-				return response.json ? response.json() : Promise.reject( response );
+			if ( ! response || ! response.json ) {
+				throw invalidJsonError;
 			}
 
-			return response;
-		};
-
-		return responsePromise
-			.then( checkStatus )
-			.then( parseResponse )
-			.catch( ( response ) => {
-				if ( ! parse ) {
-					throw response;
-				}
-
-				const invalidJsonError = {
-					code: 'invalid_json',
-					message: __( 'The response is not a valid JSON response.' ),
-				};
-
-				if ( ! response || ! response.json ) {
+			return response.json()
+				.catch( () => {
 					throw invalidJsonError;
-				}
-
-				return response.json()
-					.catch( () => {
-						throw invalidJsonError;
-					} )
-					.then( ( error ) => {
-						const unknownError = {
-							code: 'unknown_error',
-							message: __( 'An unknown error occurred.' ),
-						};
-
-						throw error || unknownError;
-					} );
-			} );
-	};
+				} )
+				.then( ( error ) => {
+					const unknownError = {
+						code: 'unknown_error',
+						message: __( 'An unknown error occurred.' ),
+					};
+
+					throw error || unknownError;
+				} );
+		} );
+};
 
-	const steps = [
-		raw,
+let fetchHandler = defaultFetchHandler;
+
+function setFetchHandler( newFetchHandler ) {
+	fetchHandler = newFetchHandler;
+}
+
+function apiFetch( options ) {
+	const builtInMiddlewares = [
 		fetchAllMiddleware,
 		httpV1Middleware,
 		namespaceEndpointMiddleware,
 		userLocaleMiddleware,
-		...middlewares,
+	];
+
+	const steps = [
```

Comment:
I'd normally not have brought it up, but since you did some refactoring here, I'm a bit curious why we construct as much as we do here on every call to `apiFetch`. Why not just initialize the scoped `middlewares` to include `fetchAllMiddleware` et al? I think by eliminating `steps` we could even make clearer that the fetch handler is the terminating behavior. Looking at `runStep`'s creation by `index + 1`, it took me more than a few seconds to understand how we'd _not_ risk some `undefined` value being called as a function.†

† In fact, while this probably _won't_ normally occur, I think an error would be thrown if the fetch handler actually called `next` ?

Reply:
```diff
diff --git a/packages/api-fetch/src/test/index.js b/packages/api-fetch/src/test/index.js
index d993a8b63..0149a9efb 100644
--- a/packages/api-fetch/src/test/index.js
+++ b/packages/api-fetch/src/test/index.js
@@ -196,5 +196,6 @@ describe( 'apiFetch', () => {
    expect( customFetchHandler ).toHaveBeenCalledWith( {
      path: '/random?_locale=user',
    }, expect.any( Function ) );
+   expect( customFetchHandler.mock.calls[ 0 ][ 1 ] ).not.toThrow();
  } );
 } );
```

```
● apiFetch › should not use the default fetch handler when using a custom fetch handler

    expect(function).not.toThrow()

    Expected the function not to throw an error.
    Instead, it threw:
      TypeError: nextStep is not a function
          136 | 		const nextStep = steps[ index ];
          137 | 		const next = runStep( index + 1 );
        > 138 | 		return nextStep( nextOptions, next );
              | 		       ^
          139 | 	};
          140 | 
          141 | 	return runStep( 0 )( options );
    
          at nextStep (packages/api-fetch/src/index.js:138:10)
          at Object.<anonymous> (node_modules/expect/build/to_throw_matchers.js:51:9)
          at Object.throwingMatcher [as toThrow] (node_modules/expect/build/index.js:320:33)
          at Object.toThrow (packages/api-fetch/src/test/index.js:199:57)

      197 | 			path: '/random?_locale=user',
      198 | 		}, expect.any( Function ) );
    > 199 | 		expect( customFetchHandler.mock.calls[ 0 ][ 1 ] ).not.toThrow();
          | 		                                                      ^
      200 | 	} );
      201 | } );
      202 | 

      at Object.toThrow (packages/api-fetch/src/test/index.js:199:57)
```

Reply:
> I'd normally not have brought it up, but since you did some refactoring here, I'm a bit curious why we construct as much as we do here on every call to `apiFetch`. Why not just initialize the scoped `middlewares` to include `fetchAllMiddleware` et al? I think by eliminating `steps` we could even make clearer that the fetch handler is the terminating behavior. Looking at `runStep`'s creation by `index + 1`, it took me more than a few seconds to understand how we'd _not_ risk some `undefined` value being called as a function.†
> 
> † In fact, while this probably _won't_ normally occur, I think an error would be thrown if the fetch handler actually called `next` ?

I pushed a few commits to sort out my concerns.